### ISSUE TITLE
Anovos_use_case_demo.ipynb shows error when generating report

### ISSUE THREAD

**Issue by ZhuLi99** [anovos/anovos#312]
The Jupyter notebook anovos_use_case_demo.ipynb stored [here](https://github.com/anovos/anovos/blob/main/tutorial/use_case_demo/anovos_use_case_demo.ipynb) is now getting many error messages when generating **Full Report**. 
## Expected Behavior
In the notebook published on Github, the **Full Report** section should look like this
![Screenshot 2022-10-04 at 3 45 10 PM](https://user-images.githubusercontent.com/112929089/193763574-5637f4c4-ad41-4445-82fc-5b78f89a3a41.png)

## Current Behavior
- When I run it using **Anovos 1.0.0**, it could successfully generate the report while **charts_to_object** function shows below error messages.
`2022-10-04 16:15:06.762 | ERROR    | anovos.data_report.report_preprocessing:edit_binRange:138 - processing failed during edit_binRange, error 'NoneType' object has no attribute 'split'`
![image](https://user-images.githubusercontent.com/112929089/193770282-15b4d017-5a29-43e7-857d-454a09ef947c.png)

(The correlation matrix error is because **anovos_basic_report** function has set **skip_corr_matrix=True** by default. This would prevent the Full Report from generating correlation matrix plot, but the sample report [here](https://github.com/anovos/anovos/blob/main/tutorial/use_case_demo/assets/reports/ml_anovos_report.html) still has the plot. This might confuse the users when running the demo). 

- When I run it using **Anovos 1.0.1**, it failed to complete the **anovos_basic_report** function due to errors in variable clustering function.

![image](https://user-images.githubusercontent.com/112929089/193771758-86d7ccfb-5df6-4401-8367-9c543c526c9a.png)

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [9], in <cell line: 5>()
      2 report_path = output_path+"data_report/report_stats/"
      4 #Option 1 (with ID & Label Column)
----> 5 anovos_basic_report(spark, df_sample, id_col='SK_ID_CURR', label_col="TARGET", 
      6                     event_label=1, output_path=report_path, print_impact=False)

File ~/opt/anaconda3/lib/python3.9/site-packages/anovos/data_report/basic_report_generation.py:186, in anovos_basic_report(spark, idf, id_col, label_col, event_label, skip_corr_matrix, output_path, run_type, auth_key, print_impact)
    184 elif func in AA_funcs:
    185     extra_args = stats_args(output_path, func.__name__)
--> 186     stats = func(spark, idf, drop_cols=id_col, **extra_args)
    187 elif label_col:
    188     if func in AT_funcs:

File ~/opt/anaconda3/lib/python3.9/site-packages/anovos/data_analyzer/association_evaluator.py:258, in variable_clustering(spark, idf, list_of_cols, drop_cols, sample_size, stats_unique, stats_mode, print_impact)
    256 idf_pd = idf_imputed.toPandas()
    257 vc = VarClusHi(idf_pd, maxeigval2=1, maxclus=None)
--> 258 vc.varclus()
    259 odf_pd = vc.rsquare
    260 odf = spark.createDataFrame(odf_pd).select(
    261     "Cluster",
    262     F.col("Variable").alias("Attribute"),
    263     F.round(F.col("RS_Ratio"), 4).alias("RS_Ratio"),
    264 )

File ~/opt/anaconda3/lib/python3.9/site-packages/varclushi/varclushi.py:215, in VarClusHi.varclus(self, speedup)
    212 self.speedup = speedup
    214 if self.speedup is True:
--> 215     return self._varclusspu()
    217 ClusInfo = collections.namedtuple('ClusInfo', ['clus','eigval1','eigval2','pc1','varprop'])
    218 c_eigvals, _, c_princomps, c_varprops = VarClusHi.pca(self.df[self.feat_list])

File ~/opt/anaconda3/lib/python3.9/site-packages/varclushi/varclushi.py:143, in VarClusHi._varclusspu(self)
    140 def _varclusspu(self):
    142     ClusInfo = collections.namedtuple('ClusInfo', ['clus', 'eigval1', 'eigval2', 'eigvecs','varprop'])
--> 143     c_eigvals, c_eigvecs, c_corrs, c_varprops = VarClusHi.correig(self.df[self.feat_list])
    145     self.corrs = c_corrs
    147     clus0 = ClusInfo(clus=self.feat_list,
    148                      eigval1=c_eigvals[0],
    149                      eigval2=c_eigvals[1],
    150                      eigvecs=c_eigvecs,
    151                      varprop=c_varprops[0]
    152                      )

File ~/opt/anaconda3/lib/python3.9/site-packages/varclushi/varclushi.py:38, in VarClusHi.correig(df, feat_list, n_pcs)
     36     varprops = [sum(eigvals)]
     37 else:
---> 38     corr = np.corrcoef(df.values.T)
     39     raw_eigvals, raw_eigvecs = np.linalg.eigh(corr)
     40     idx = np.argsort(raw_eigvals)[::-1]

File <__array_function__ internals>:5, in corrcoef(*args, **kwargs)

File ~/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/function_base.py:2683, in corrcoef(x, y, rowvar, bias, ddof, dtype)
   2679 if bias is not np._NoValue or ddof is not np._NoValue:
   2680     # 2015-03-15, 1.10
   2681     warnings.warn('bias and ddof have no effect and are deprecated',
   2682                   DeprecationWarning, stacklevel=3)
-> 2683 c = cov(x, y, rowvar, dtype=dtype)
   2684 try:
   2685     d = diag(c)

File <__array_function__ internals>:5, in cov(*args, **kwargs)

File ~/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/function_base.py:2518, in cov(m, y, rowvar, bias, ddof, fweights, aweights, dtype)
   2515     else:
   2516         w *= aweights
-> 2518 avg, w_sum = average(X, axis=1, weights=w, returned=True)
   2519 w_sum = w_sum[0]
   2521 # Determine the normalization

File <__array_function__ internals>:5, in average(*args, **kwargs)

File ~/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/function_base.py:380, in average(a, axis, weights, returned)
    377 a = np.asanyarray(a)
    379 if weights is None:
--> 380     avg = a.mean(axis)
    381     scl = avg.dtype.type(a.size/avg.size)
    382 else:

File ~/opt/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:181, in _mean(a, axis, dtype, out, keepdims, where)
    179 ret = umr_sum(arr, axis, dtype, out, keepdims, where=where)
    180 if isinstance(ret, mu.ndarray):
--> 181     ret = um.true_divide(
    182             ret, rcount, out=ret, casting='unsafe', subok=False)
    183     if is_float16_result and out is None:
    184         ret = arr.dtype.type(ret)
TypeError: unsupported operand type(s) for /: 'str' and 'int'
```

## Steps to reproduce
1. Download the use case demo jupyter notebook [here](https://github.com/anovos/anovos/blob/main/tutorial/use_case_demo/anovos_use_case_demo.ipynb)
2. Run the commands until **Full Report** section and errors will show up depending on which Anovos version we are using

### Specifications

- Version: Anovos 1.0.0 / Anovos 1.0.1
- Platform: MacOS, M1 Chip
- Subsystem: Python 3.9, Spark 3.3.0

## Possible Solution
- For errors using Anovos 1.0.0, the errors related to splitting NoneType objects was shown up during **charts_to_object** function call. This might be due to passing NoneType to edit_binRange function when generating some plots.

**Comment by ionicsolutions** [comment]
PR #311 contains a demonstration of the issue using a workflow configuration

The associated tests (https://github.com/anovos/anovos/blob/main/src/test/anovos/data_analyzer/test_association_evaluator.py#L264) apparently do not capture the case. There, columns are dropped before being passed into `variable_clustering()` but it seems that otherwise the same dataset is used.

As a next step, the error should be reproduced using a minimal example, preferably by passing a hand-crafted Pandas DataFrame when instantiating `VarClushHi()`.

**Comment by VishnuGowthem** [comment]
once #311 is completed and merged, I think the issue in variable clustering issue should be handled.
Also, we haven't updated the tutorial/use case image after v0.3.0 release.
Hence for now, I updated the anovos/anovos-workshop image on Dockerhub manually (Note : we don't have an automated workflow for that yet)
Documentation for the same is done  #313.

**Comment by ionicsolutions** [comment]
@VishnuGowthem #311 is just a demonstration/isolation of the issue. We should discuss today what the next steps are.

**Comment by ZhuLi99** [comment]
I have some findings for the bug on Variable-Clustering. 

The error is observed in variable_clustering function which we did not do any updates in previous release. The logic for variable clustering is encode input data --> imputation --> initiate VarClusHi object --> run variable clustering commands. The actual error is raised in the encoding part, caused by initiating  VarClusHi class with a dataframe containing **non-numerical** columns.
When VarClusHi runs variable clustering commands, it need to calculate the correlation between values first using this command `corr = np.corrcoef(df.values.T)`. Since the dataframe contains categorical columns, it failed to calculate correlation and give errors.

We have updated cat_to_num_unsupervised on [Sep 15th](https://github.com/anovos/anovos/pull/247) and updated the logic for [cat_to_num_unsupervised](https://github.com/anovos/anovos/pull/247/commits/9a559c442df096b9f5db9a09ba6f53c29802b09b) function. 

In previous logic
- cardinality_threshold is used only when method_type = "one-hot encoding", and those columns with uniqueCount > cardinality_threshold will be treated as skip_cols.
```
skip_cols = []
if method_type == "onehot_encoding":
    if stats_unique == {}:
        skip_cols = (
            uniqueCount_computation(spark, idf, list_of_cols)
            .where(F.col("unique_values") > cardinality_threshold)
            .select("attribute")
            .rdd.flatMap(lambda x: x)
            .collect()
        )
    else:
        skip_cols = (
            read_dataset(spark, **stats_unique)
            .where(F.col("unique_values") > cardinality_threshold)
            .select("attribute")
            .rdd.flatMap(lambda x: x)
            .collect()
        )
    skip_cols = list(
        set([e for e in skip_cols if e in list_of_cols and e not in drop_cols])
    )

    if skip_cols:
        warnings.warn(
            "Columns dropped from one-hot encoding due to high cardinality: "
            + ",".join(skip_cols)
        )
```
- But the default setting for method_type is 1 (label_encoding). In other words, if no method_type was specified when calling cat_to_num_unsupervised, all categorical columns will be label-encoded. This means the output data would not contain any string columns.

In update logic
- Cardinality_threshold is checked regardless of method_type. This means that for columns with uniqueCount > cardinality_threshold, it will not be encoded.
```
if stats_unique == {}:
    skip_cols = (
        uniqueCount_computation(spark, idf, list_of_cols)
        .where(F.col("unique_values") > cardinality_threshold)
        .select("attribute")
        .rdd.flatMap(lambda x: x)
        .collect()
    )
else:
    skip_cols = (
        read_dataset(spark, **stats_unique)
        .where(F.col("unique_values") > cardinality_threshold)
        .select("attribute")
        .rdd.flatMap(lambda x: x)
        .collect()
    )
skip_cols = list(
    set([e for e in skip_cols if e in list_of_cols and e not in drop_cols])
)

if skip_cols:
    warnings.warn(
        "Columns dropped from encoding due to high cardinality: "
        + ",".join(skip_cols)
    )
```
- For example, the "ifa" column in income_dataset will not be encoded under new logic with label encoding (as it has 32,561 unique values) while in previous logic "ifa" will be encoded. We can see the clustering result [here](https://github.com/anovos/anovos/blob/main/examples/notebooks/data_analyzer__association_evaluator.ipynb) and it shows "ifa" is in a separate cluster with RS_ratio = 0.

**Comment by ZhuLi99** [comment]
My solution is to do an extra attribute-type check after encoding, and only select the numerical columns before initiating VarClusHi.
```
num_cols, cat_cols, other_cols = attributeType_segregation(idf_encoded)
idf_encoded = idf_encoded.select(num_cols)
```

I have tested this logic locally under geospatial branch and could solve the problem. I have pushed the commit on Geospatial_Combined branch and will give update on the results after all checks have been completed.
